### PR TITLE
feat: add elasticsearch for datasets finder

### PR DESCRIPTION
**Pull request by samuelhwilliams**
### Checklist

* [ ] Have tests been added to cover any changes?

### CODE REVIEW

File: infra/security_groups.tf [hunk 1]
```diff
@@ -1501,3 +1501,27 @@ resource "aws_security_group_rule" "datasets_db_egress_all_to_quicksight" {
...
+  from_port   = "443"
+  to_port     = "443"
+  protocol    = "tcp"
+}
+
+resource "aws_security_group_rule" "elasticsearch_ingress_from_paas" {
+  description = "ingress-elasticsearch-https-from-paas-ie-data-flow"
+
+  security_group_id = "${aws_security_group.datasets.id}"
+  cidr_blocks       = ["10.0.0.0/16"]
```
Comment (michalc) [inline infra/security_groups.tf:1521]: It's unlikely to change in any rush, but should the terraform find `10.0.0.0/16` at runtime from `aws_vpc_peering_connection.datasets_to_paas` for clarity?

File: infra/elasticsearch_datasets.tf [hunk 2]
```diff
@@ -0,0 +1,84 @@
+resource "aws_iam_service_linked_role" "datasets_finder" {
+  # This is a shared resource between all envs ... couldn't find a way to have
+  # one per env as it automatically assigns the same name (and custom suffixes aren't
+  # allowed for ES service linked roles).
+  aws_service_name = "es.amazonaws.com"
+}
+
+resource "aws_elasticsearch_domain" "datasets_finder" {
+  domain_name           = var.prefix
```
Comment (michalc) [inline infra/elasticsearch_datasets.tf:9]: Can/should this include something like "dataset-finder", just for clarity on the AWS-side?

File: infra/elasticsearch_datasets.tf [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  access_policies = <<CONFIG
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Principal": {
+        "AWS": "${aws_iam_role.admin_task.arn}"
+      },
+      "Action": "es:*",
```
Comment (michalc) [inline infra/elasticsearch_datasets.tf:51]: Can this be locked down to be read only?

File: infra/elasticsearch_datasets.tf [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+      "Resource": "arn:aws:es:${data.aws_region.aws_region.name}:${data.aws_caller_identity.aws_caller_identity.account_id}:domain/${var.prefix}/*"
+    },
+    {
+      "Effect": "Allow",
+      "Principal": {
+        "AWS": "${aws_iam_user.datasets_finder_data_flow.arn}"
+      },
+      "Action": "es:*",
+      "Resource": "arn:aws:es:${data.aws_region.aws_region.name}:${data.aws_caller_identity.aws_caller_identity.account_id}:domain/${var.prefix}/*"
+    }
```
Comment (michalc) [inline infra/elasticsearch_datasets.tf:61]: Erring on the side of paranoia, can the data_flow one be restricted to the PaaS IP range, and the admin one to the IP range of the VPC its in?

File: infra/security_groups.tf (hunk 1 above)
Comment (samuelhwilliams) [reply infra/security_groups.tf:1521]: I think we'd need to make that a terraform data source to retrieve - so going to use `var.paas_cidr_block` instead.

File: infra/elasticsearch_datasets.tf (hunk 2 above)
Comment (samuelhwilliams) [reply infra/elasticsearch_datasets.tf:9]: 👍

File: infra/elasticsearch_datasets.tf (hunk 3 above)
Comment (samuelhwilliams) [reply infra/elasticsearch_datasets.tf:51]: 👍

File: infra/elasticsearch_datasets.tf (hunk 4 above)
Comment (samuelhwilliams) [reply infra/elasticsearch_datasets.tf:61]: 👍

**Comment by samuelhwilliams** [comment]
@michalc good for re-review .. flaky tests are flaky.

**Review by michalc — APPROVED** [review]

File: infra/elasticsearch_datasets.tf (hunk 4 above)
Comment (samuelhwilliams) [reply infra/elasticsearch_datasets.tf:75]: Actually, looks like AWS doesn't let you apply IP based policies to a VPC endpoint ES instance...

Comment (samuelhwilliams) [reply infra/elasticsearch_datasets.tf:75]: I think probably fine to go without - IP controls are in place on the security group to limit access?

**Pull request merged**